include/boost/math/concepts/concepts.hpp

```diff
@@ -0,0 +1,111 @@
+//  (C) Copyright Matt Borland 2022.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_CONCEPTS_CONCEPTS_HPP
+#define BOOST_MATH_CONCEPTS_CONCEPTS_HPP
+
+#if __cplusplus >= 202002L || _MSVC_LANG >= 202002L
+#if __has_include(<concepts>)
+#include <concepts>
+
+#include <type_traits>
+#include <boost/math/tools/config.hpp>
+
+namespace boost::math::concepts {
+
+template <typename T>
+concept Integral = std::is_integral_v<T>
+                   #ifdef __SIZEOF_INT128__
+                   || std::is_same_v<__int128_t, T>
+                   || std::is_same_v<__uint128_t, T>
+                   #endif
+                   ;
+
+
+template <typename T>
+concept Signed_integral = std::is_integral_v<T> && std::is_signed_v<T>
+                          #ifdef __SIZEOF_INT128__
+                          || std::is_same_v<__int128_t, T>
+                          #endif
+                          ;
+
+template <typename T>
+concept Unsigned_integral = std::is_integral_v<T> && std::is_unsigned_v<T>
+                            #ifdef __SIZEOF_INT128__
+                            || std::is_same_v<__uint128_t, T>
+                            #endif
+                            ;
+
+template <typename T>
+concept Real = std::is_floating_point_v<T>
+               #ifdef BOOST_HAS_FLOAT128
+               || std::is_same_v<__float128, T>
```

Comment:
Would this work with `MPFR`/ `multiprecision` types?

Reply:
Same question as with integral types